src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPDocumentSynchronizer.cs

```diff
@@ -94,45 +135,77 @@ internal void DocumentManager_Changed(object sender, LSPDocumentChangeEventArgs
                 throw new ArgumentNullException(nameof(args));
             }
 
-            if (args.Kind != LSPDocumentChangeKind.VirtualDocumentChanged)
-            {
-                return;
-            }
-
-            var lspDocument = args.New;
-            for (var i = 0; i < lspDocument.VirtualDocuments.Count; i++)
+            lock (DocumentContextLock)
             {
-                if (!_synchronizingContexts.TryGetValue(lspDocument.VirtualDocuments[i].Uri, out var synchronizingContext))
+                if (args.Kind == LSPDocumentChangeKind.Added)
                 {
-                    continue;
-                }
+                    var lspDocument = args.New;
+                    for (var i = 0; i < lspDocument.VirtualDocuments.Count; i++)
+                    {
+                        var virtualDocument = lspDocument.VirtualDocuments[i];
 
-                if (lspDocument.Version == synchronizingContext.ExpectedHostDocumentVersion)
-                {
-                    synchronizingContext.SetSynchronized(true);
+                        Debug.Assert(!_virtualDocumentContexts.ContainsKey(virtualDocument.Uri));
+
+                        var virtualDocumentTextBuffer = virtualDocument.Snapshot.TextBuffer;
+                        virtualDocumentTextBuffer.PostChanged += VirtualDocumentBuffer_PostChanged;
+                        _virtualDocumentContexts[virtualDocument.Uri] = new DocumentContext(_synchronizationTimeout);
+                    }
                 }
-                else if (lspDocument.Version > synchronizingContext.ExpectedHostDocumentVersion)
+                else if (args.Kind == LSPDocumentChangeKind.Removed)
                 {
-                    // The LSP document version has surpassed what the projected document was expecting for a version. No longer able to synchronize.
-                    synchronizingContext.SetSynchronized(false);
+                    var lspDocument = args.Old;
+                    for (var i = 0; i < lspDocument.VirtualDocuments.Count; i++)
+                    {
+                        var virtualDocument = lspDocument.VirtualDocuments[i];
+
+                        Debug.Assert(_virtualDocumentContexts.ContainsKey(virtualDocument.Uri));
+
+                        var virtualDocumentTextBuffer = virtualDocument.Snapshot.TextBuffer;
```

Comment:
My instinct is that a `TextBuffer` needs to be disposed when we're done with it. Am I off the mark, or maybe that's taken care of elsewhere?

Reply:
Off the mark. `TextBuffers` are VS' utility to track text for documents. VS takes care of creating them and then they go away once all references to them go away.

Basically for every document in VS (visible or non-visible) there's a TextBuffer backing that document.